upgradeability_using_inherited_storage/contracts/Registry.sol

```diff
@@ -1,24 +1,24 @@
 pragma solidity ^0.4.18;
 
 import './IRegistry.sol';
-import './Proxy.sol';
 import './Upgradeable.sol';
+import './UpgradeabilityProxy.sol';
 
 contract Registry is IRegistry {
   mapping (string => address) versions;
 
-  function addVersion(string version, address impl) public {
+  function addVersion(string version, address implementation) public {
     require(versions[version] == 0x0);
-    versions[version] = impl;
-    VersionAdded(version, impl);
+    versions[version] = implementation;
+    VersionAdded(version, implementation);
   }
 
   function getVersion(string version) public view returns (address) {
     return versions[version];
   }
 
-  function create(string version) public payable returns (Proxy) {
-    Proxy proxy = new Proxy(version);
+  function create(string version) public payable returns (UpgradeabilityProxy) {
+    UpgradeabilityProxy proxy = new UpgradeabilityProxy(version);
 
     Upgradeable(proxy).initialize.value(msg.value)(msg.sender);
```

Comment:
why the cast to `Upgradeable` here?

Reply:
Because we need to call the `initialize` function that is defined in the `Upgradeable`, which is the contract that every implementation of the upgradeable behavior will need. If we don't perform that cast, the compiler will throw a `TypeError` error since `initialize` is not defined in the `UpgradeabilityProxy`.